conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/RequestContext.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.lib;
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.ListMultimap;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Interface providing a view over data provided by the the original HTTP request including request headers and query
+ * parameters. This is a read only interface which should only be implemented by {@code conjure-java-undertow-runtime}.
+ */
+@Beta
+public interface RequestContext {
+
+    /**
+     * Returns all values of the header named {@code headerName}. The name is case insensitive. An empty list is
+     * returned if no such header exists.
+     */
+    List<String> header(String headerName);
+
+    /**
+     * Returns the value of the first header named {@code headerName}. The name is case insensitive.
+     * An {@link Optional#empty()} is returned if no such header exists.
+     */
+    Optional<String> firstHeader(String headerName);
+
+    /**
+     * Returns all query parameters associated with the current request.
+     */
+    ListMultimap<String, String> queryParameters();
```

Comment:
Why not just Multimap? The underlying is a Map<String, Dequeue<String>>, doesn't feel like you gain anything from list and you'd be able to maybe construct it more cheaply?

Reply:
I'm not sure there's a multimap wrapper type we could use, I'd rather use a strongly typed `ListMultimap` than `Multimap` to avoid forcing consumers to interact with `Collection`. In this case I'm not concerned with performance because the immutable maps is only created when it's requested, which should be rare. Definitely want to avoid `SetMultimap` because we don't want to hash values, or build trees.